### PR TITLE
♻️  🏗  Update build-system to build bento components stored in a new directory

### PR DESCRIPTION
**Pull request by rileyajones**
The first step in moving bento to its own repository is to separate bento from the rest of the repository.

Create new `/bento` directory, add support for building components stored within it to the build system and move over a bento component.

After this initial move I will submit a series of PRs moving over the remaining Bento components and potentially renaming them from `amp-*` to `bento-*`.

#36461

### CODE REVIEW

**Review by rcebulko — COMMENTED** [review]
WDYT about instead creating `bento/` within `src/` and granting it the alias `#bento`? This would be in line with https://github.com/ampproject/amphtml/issues/34866, which would eventually see `extensions` itself moving under `src` as well (but this is difficult to do without disrupting development, and so hasn't happened yet), but we've already been moving some others there. The eventual idea would be for `src/` to contain all JS that gets shipped in binaries, and everything else in the top-level directory serves some other purpose (ie. `build-system`, `docs`, `examples`, dotfiles, etc.

As a side note, this also puts us one eslint config tweak away from restricting imports within `#bento` using our existing `no-restricted-imports` rule currently in place within `src`

**Comment by rcebulko** [comment]
Also, because I realize I got distracted by my other comment, I forgot to mention: This is awesome, thank you for doing this! Moving big chunks like this is effortful but sorely needed imo

File: build-system/compile/bundles.config.js
Comment (samouri) [inline build-system/compile/bundles.config.js:5]:
```suggestion
```

This duped with line 213

File: build-system/babel-config/test-config.js
Comment (estherkim) [inline build-system/babel-config/test-config.js:20]: should start with `src/`?

File: build-system/compile/bundles.config.bento.json
Comment (estherkim) [inline build-system/compile/bundles.config.bento.json:9]: maybe these will be redundant in the future since this is a specific json for bento

File: build-system/compile/bundles.config.js
Comment (estherkim) [inline build-system/compile/bundles.config.js:5]: dupe of line 213?

File: build-system/compile/sources.js
Comment (estherkim) [inline build-system/compile/sources.js:66]: start with `src/`?

File: build-system/tasks/build-bento.js
Comment (estherkim) [inline build-system/tasks/build-bento.js:258]: nit: rename to `buildBentoComponents`

**Review by samouri — COMMENTED** [review]
What do we want to do about AMP Components vs. Standalone WC/Preact/React Distributables?

My take is that the new build-system path you create should only be responsible for building the: Standalone WC / Preact / React components. _Not_ the AMP-layer.

Eventually if Bento moved to its own Repo, the AMP Component would not come along for the ride.

File: build-system/compile/bundles.config.js
Comment (samouri) [reply build-system/compile/bundles.config.js:5]: Jinx!

Comment (rileyajones) [reply build-system/compile/bundles.config.js:5]: Good catch! Not sure how that happened, maybe a rebase issue?

File: build-system/compile/sources.js
Comment (rileyajones) [reply build-system/compile/sources.js:66]: 👍

File: src/bento/components/amp-jwplayer/1.0/base-element.js
Comment (samouri) [inline src/bento/components/amp-jwplayer/1.0/base-element.js:4]: To make my question above more concrete, if we decide to move only the WC/Preact/React layers, then I'd expect the `amp-jwplayer.js` file to stay where it is today

File: build-system/tasks/build-bento.js
Comment (rileyajones) [reply build-system/tasks/build-bento.js:197]: That's great to know! Thanks

File: build-system/tasks/build-bento.js
Comment (samouri) [inline build-system/tasks/build-bento.js:198]:
```suggestion
```

File: build-system/tasks/build-bento.js
Comment (samouri) [inline build-system/tasks/build-bento.js:209]:
```suggestion
```

File: build-system/tasks/build-bento.js
Comment (samouri) [inline build-system/tasks/build-bento.js:239]:
```suggestion
  let o = {...options, ...e};
```

File: build-system/tasks/build-bento.js
Comment (samouri) [inline build-system/tasks/build-bento.js:259]: I find this shadowing relatively confusing. There is a var named `componentsToBuild` on the module-scope.
Can we get rid of that? (module-scope variables are a code smell IMO)

File: build-system/tasks/build-bento.js
Comment (rcebulko) [inline build-system/tasks/build-bento.js:92]: For your consideration:
```suggestion
    const allComponents = Object.values(components).map(c => c.name);
```

File: build-system/tasks/build-bento.js
Comment (rcebulko) [inline build-system/tasks/build-bento.js:111]: typo

File: build-system/tasks/build-bento.js
Comment (rcebulko) [inline build-system/tasks/build-bento.js:190]:
```suggestion
    mkdirSync('build/css', {recursive: true});
```

File: build-system/tasks/build-bento.js
Comment (rcebulko) [inline build-system/tasks/build-bento.js:208]: This function conditionally kicks off a decent number of async tasks and awaits each one in sequence. If they aren't order-dependent, it might make sense to create a `tasks = []` list and directly append each promise to it in the loop, then `return Promise.all(tasks)`.

File: build-system/tasks/build-bento.js
Comment (rcebulko) [inline build-system/tasks/build-bento.js:238]: `buildComponent`/`watchComponent` take a lot of similar args  where order is important. If they continue to grow, it may make sense to move more/all of them into a single options object, which would simplify some of this code that mostly forwards options along

File: build-system/tasks/build-bento.js
Comment (rcebulko) [inline build-system/tasks/build-bento.js:253]: I think I may be missing something; where does `components` come from here?

File: build-system/tasks/build-bento.js
Comment (rcebulko) [inline build-system/tasks/build-bento.js:261]: It feels odd to me that `doBuildComponent` takes an object and a key, then immediately does `components[component]` with it; likewise, we're looping over keys here, then indexing. If we instead loop over the values directly and pass them directly, with `doBuildComponent accepting the component object directly, I think it becomes easier to follow.

Feel free to ignore the `map`/functional approach, but either way it seems like looping over/passing the component objects directly could be cleaner here

```suggestion
  await Promise.all(
    Object.values(components)
      .filter(c => options.compileOnlyCss || toBuild.includes(c.name))
      .map(c => doBuildComponent(c, options))
  );
```

File: build-system/tasks/build.js
Comment (rcebulko) [inline build-system/tasks/build.js:54]: nit: consider `await Promise.all([buildExtensions(options), buildBentoComponents(options)]);`
I'm not sure how efficiently the rest of our build systems, esbuild, closure, w/e handle running in parallel but at least this way if they do (now or eventually) we can take better advantage of it.

File: build-system/tasks/css/index.js
Comment (rcebulko) [inline build-system/tasks/css/index.js:143]:
```suggestion
  await Promise.all([buildExtensions(buildOptions), buildBentoComponents(buildOptions)]);
```

File: build-system/tasks/dist.js
Comment (rcebulko) [inline build-system/tasks/dist.js:136]: nit: consider. `Promise.all`

File: src/bento/.eslintrc.js
Comment (rcebulko) [inline src/bento/.eslintrc.js:6]: nit: `import/order` is enabled for everything in `src/`, so you only need the exclusion override here
```suggestion
```

File: src/bento/components/amp-jwplayer/1.0/base-element.js
Comment (rcebulko) [reply src/bento/components/amp-jwplayer/1.0/base-element.js:4]: If I'm understanding @samouri correctly, I agree with this sentiment, as it fits more in line with "AMP is a downstream user of Bento"

**Review by rcebulko — COMMENTED** [review]
👏🏻 👏🏻 👏🏻

File: .eslintrc.js
Comment (rileyajones) [reply .eslintrc.js:463]: I intend to move `src/preact` into `src/bento` as well because it will also need to move to the new repo.

File: src/bento/.eslintrc.js
Comment (rileyajones) [reply src/bento/.eslintrc.js:6]: Good catch, I missed this when I moved `/bento` to `src/bento`

File: build-system/tasks/build.js
Comment (rileyajones) [reply build-system/tasks/build.js:54]: I am also not sure how efficient parallel closure builds will be but it seems worth checking. If this were intend to stay for a long time I would have each of these return a set of files to be built which would then be passed to a single invocation of Closure

File: build-system/tasks/build-bento.js
Comment (rileyajones) [reply build-system/tasks/build-bento.js:195]: TBH this whole file is mostly a tweaked version of `build-system/tasks/extension-helpers.js` and I was attempting to change as little as possible in this initial version. There is a lot strange kinda funky code there that I would like to clean up in a future iteration.

File: .eslintrc.js
Comment (rcebulko) [reply .eslintrc.js]: So in that case `src/bento` would still cover it?

Comment (rileyajones) [reply .eslintrc.js]: Yes, though given that `src/preact` is also in this list that is probably not an actual issue. I'll go ahead and loosen this up.

File: build-system/tasks/build.js
Comment (samouri) [reply build-system/tasks/build.js]: for bento in particular there is no reason to begin with closure

File: build-system/compile/bundles.config.bento.json
Comment (samouri) [inline build-system/compile/bundles.config.bento.json:6]: For the purposes of Bento configuration json, does it make sense to remove `latestVersion`, and the entire options obj? `npm` and `bento` will always be true

Comment (rileyajones) [reply build-system/compile/bundles.config.bento.json]: Yeah, that makes sense. I'll remove it from the config but stub it in JS for now because there are shared parts of the build process that check those options.

File: build-system/tasks/build-bento.js
Comment (rileyajones) [reply build-system/tasks/build-bento.js:214]: Disclaimer: This code is essentially a replica of that found in `build-system/tasks/extension-helpers.js` so while I can explain HOW it works, I really don't understand WHY it is written this way. It seems to be to prevent repeat calls to the method from duplicating work.

`components` is declared globally on line 24 and is then populated by the method `declareExctension` called by `maybeInitializeComponents`

**Comment by samouri** [comment]
@rileyajones: I'm unclear right now if this PR is meant to be reviewed or do you mean to put it back into "draft" state?

File: src/bento/components/amp-jwplayer/1.0/component.type.js
Comment (samouri) [inline src/bento/components/amp-jwplayer/1.0/component.type.js:6]: Question for Bento developers: are there any attributes that do not work on the CE standalone component and only work on the AMP component? If so, we'll need to create two type files

File: src/bento/components/amp-jwplayer/OWNERS
Comment (rcebulko) [reply src/bento/components/amp-jwplayer/OWNERS:7]: `pablos1` was an existing owner in the file; I'm in favor of leaving it be in this PR

File: build-system/tasks/build-bento.js
Comment (rcebulko) [inline build-system/tasks/build-bento.js:67]: Almost every usage I've found of `dedupe` follows this pattern: `dedupe(arr1.concat(arr2))` or `dedupe([...arr1, ...arr2])`
It's currently implemented as
```

/**
 * Remove duplicates from the given array.
 * @param {!Array<string>} arr
 * @return {!Array<string>}
 */
function dedupe(arr) {
  const map = Object.create(null);
  arr.forEach((item) => (map[item] = true));
  return Object.keys(map);
}
```

It could be
```
/**
 * Remove duplicates from the given array.
 * @param {!Array<string>...} arrs
 * @return {!Array<string>}
 */
function dedupe(arrs...) {
  const s = new Set();
  arrs.forEach(arr => s.add(...arr));
  return s.values();
}
```

Still doing pretty much the same thing, but would make for simple call-sites

Comment (rcebulko) [reply build-system/tasks/build-bento.js:67]: As a totally separate note, have you considered making `componentsToBuild` itself a `Set`? We repeatedly concat+dedupe here (albeit in mutually exclusive `if`s IIUC), a set might be more semantic

File: build-system/tasks/build-bento.js
Comment (rcebulko) [inline build-system/tasks/build-bento.js:155]: `mkdir`/`mkdirSync` don't need the `exists` check when `recursive` is used
```suggestion
    mkdirSync('build/css', {recursive: true});
```

File: build-system/tasks/build-bento.js
Comment (rcebulko) [inline build-system/tasks/build-bento.js:159]: nit: one less async/await promise wrapper
```suggestion
      return Promise.all(promises);
```

File: build-system/tasks/build-bento.js
Comment (rcebulko) [inline build-system/tasks/build-bento.js:170]: nit: one less async/await promise wrapper
```suggestion
    return Promise.all(promises);
```

File: src/bento/components/amp-jwplayer/1.0/base-element.js
Comment (rileyajones) [reply src/bento/components/amp-jwplayer/1.0/base-element.js:8]: Yeah... this is far from optimal but this should be resolved once `video-base-element` is moved into the bento directory.

File: build-system/tasks/build-bento.js
Comment (rileyajones) [reply build-system/tasks/build-bento.js:24]: This is essentially a global cache of the components that have been built. It is populated by `maybeInitializeComponents`

File: build-system/tasks/build-bento.js
Comment (rileyajones) [reply build-system/tasks/build-bento.js]: Yeah.... this should just never be null.
This code is all just a modified version of `extension-helpers` so there is a good amount of jank here.

File: build-system/tasks/build-bento.js
Comment (rileyajones) [reply build-system/tasks/build-bento.js]: As to why this is written this way, I didn't really write this so I can't really say. However, using a set definitely makes sense here.

File: build-system/tasks/build-bento.js
Comment (samouri) [reply build-system/tasks/build-bento.js:57]: Maybe I missed something, but if you have a specific (non-standard) strategy in mind for how you'd like to make incremental progress you should communicate it as early as possible! The default for reviewers is to assume new code deserves scrutiny and should start out clean.

Given your statement above it seems like your intention was to copy/paste the current code, remove what isn't needed, and then refactor in future PRs? It is likely all of the reviewers here would be ok with that plan, but we'd need to know it!

Comment (rileyajones) [reply build-system/tasks/build-bento.js:57]: Yeah copying and refactoring is the plan. After my previous comment I actually went ahead and updated the lazy builder to handle this case as well.

File: build-system/compile/bundles.config.bento.json
Comment (rileyajones) [reply build-system/compile/bundles.config.bento.json:3]: I would like to move all the bento components first, THEN rename them incase there are any issues with renaming. WDYT?

File: src/bento/components/amp-jwplayer/amp-jwplayer.md
Comment (rileyajones) [reply src/bento/components/amp-jwplayer/amp-jwplayer.md:2]: I agree with this. The documentation is copied rather than moved.

File: build-system/tasks/build-bento.js
Comment (rileyajones) [reply build-system/tasks/build-bento.js:191]: Correct, this builds `bento-jwplayer-1.0.js`

File: build-system/compile/bundles.config.bento.json
Comment (rileyajones) [reply build-system/compile/bundles.config.bento.json]: I've gone ahead and renamed this `jwplayer`

File: build-system/compile/bundles.config.bento.json
Comment (kristoferbaxter) [inline build-system/compile/bundles.config.bento.json:3]: Will the `bentojs.dev` repository need to know about this?

**Review by kristoferbaxter — COMMENTED** [review]
Looks like you have plenty of other reviewers.

File: build-system/compile/bundles.config.bento.json
Comment (rileyajones) [reply build-system/compile/bundles.config.bento.json:3]: Yes, this is going to have to be part of the build process for the new repo

File: .eslintrc.js
Comment (samouri) [inline .eslintrc.js:466]: nit: sort.  order rn seems random

File: build-system/babel-config/test-config.js
Comment (samouri) [inline build-system/babel-config/test-config.js:20]: nit: same sorting comment. this was lexicogrpahical before this insertion

File: build-system/compile/bundles.config.bento.json
Comment (samouri) [reply build-system/compile/bundles.config.bento.json:3]: I don't think Kris was talking about the new repo, he was talking about the current documentation repo: https://github.com/ampproject/bento.dev, hosted at http://bentojs.dev

File: build-system/compile/bundles.config.extensions.json
Comment (samouri) [inline build-system/compile/bundles.config.extensions.json:593]: For future PRs: we should also be able to get rid of hasCss

File: build-system/compile/bundles.config.js
Comment (samouri) [inline build-system/compile/bundles.config.js:290]: nit: for situations in which there are non-obvious positional parameters being passed around, we usually give the param name in comments. Like this:

```
verifyBundle_(
 /* condition */ ...,
 /* field */ ...,
 /* name */ ...,
 /* found */ ... 
)
```

That said, the entire function could be better inlined since imo this is hard to follow and the function is a one-liner. 🤷

File: build-system/server/lazy-build.js
Comment (samouri) [inline build-system/server/lazy-build.js:25]: nit: unclear the distinction between "extension" and "component". The pattern you followed elsewhere has also been to use the word "bento" as a replacement for "extension"

```suggestion
const bentoBundles = {};
```

File: src/bento/components/jwplayer/amp-jwplayer.md
Comment (samouri) [inline src/bento/components/jwplayer/amp-jwplayer.md:1]: For future PRs: remove all mentions of 'amp'. Rename this to README.md

**Review by samouri — APPROVED** [review]
Lots of nits. Feel free to iterate in future PRs if you'd prefer

File: build-system/compile/bundles.config.js
Comment (rileyajones) [reply build-system/compile/bundles.config.js:290]: Yeah... this function took me a bit to understand and I even thought about just rewriting it. but I wasn't sure that change belonged in this PR. I'd like to make this part of a future cleanup.

File: build-system/server/lazy-build.js
Comment (rileyajones) [reply build-system/server/lazy-build.js]: That does seem like a better name, however, I have been using the word component to refer to bento components so while this really isn't inconsistent, I will go ahead and clarify in the other situations.

File: build-system/tasks/build-bento.js
Comment (rileyajones) [reply build-system/tasks/build-bento.js:22]: I thought the uppercase name made it clearer that it was a global variable. I really hate the way that this works though and intend to revisit it in the future.

File: build-system/tasks/build-bento.js
Comment (rileyajones) [reply build-system/tasks/build-bento.js]: 🤦 missed it, thanks for pointing this out.

File: build-system/compile/bundles.config.bento.json
Comment (rileyajones) [reply build-system/compile/bundles.config.bento.json:3]: TIL there is a bento.dev repository. Yes, it looks like this needs to be updated.

BTW the linked repositories README links to [https://bento.dev/](https://bento.dev/) which seems to be an unrelated project [pr/136](https://github.com/ampproject/bento.dev/pull/136)

File: build-system/compile/bundles.config.js
Comment (rsimha) [reply build-system/compile/bundles.config.js:290]: I'm okay with rewriting / simplifying the entire verification logic. It was written at a time when a bug introduced in an extension bundle specification could silently break the build. Separate PR and lower priority, for sure. 👍

**Review by rsimha — COMMENTED** [review]
Thanks for working on this. High-level LGTM from my end, will defer to other reviewers for final approval.

**Review by rcebulko — APPROVED** [review]
A lot of this code definitely could benefit from some sprucing up, but the key changes here LGTM

**Review by estherkim — APPROVED** [review]

**Review by rebeccanthomas — APPROVED** [review]

File: src/bento/.eslintrc.js
Comment (rsimha) [inline src/bento/.eslintrc.js:1]: Curious about how we plan to maintain this file long term given that this list can get outdated pretty quickly. Okay to address separately.

**Review by rsimha — APPROVED** [review]
Thanks for working on this PR with all its moving parts. LGTM to unblock progress.

File: src/bento/.eslintrc.js
Comment (samouri) [reply src/bento/.eslintrc.js:1]: I've held the opinion for a long time that we should get rid of the file, update everything, make folks deal with merge conflicts. cc @rcebulko

Comment (rcebulko) [reply src/bento/.eslintrc.js:1]: I support this as well, teams have had plenty of time to adjust at this point

Comment (samouri) [reply src/bento/.eslintrc.js:1]: Note: I removed this file

Comment (rileyajones) [reply src/bento/.eslintrc.js:1]: That's awesome news! I'll go ahead and delete this file then.

**Comment by kvchari** [comment]
Drive-by nit: I actually like the term "extension" (as opposed to "component") because some extensions are actually composed of multiple different components (e.g. inline-gallery also exports inline-gallery-pagination/inline-gallery-thumbnails web-components, accordion only have 1 web-component but it does expose multiple preact components for section, header, and content, etc.). Are we moving away from that terminology with your changes to the directory naming and build-system changes?

**Comment by samouri** [comment]
Congratulations @rileyajones on landing this! :tada

**Comment by rileyajones** [comment]
@samouri thank you Jake!

**Pull request merged**
